### ISSUE TITLE
Alternative to workspacer bar?

### ISSUE THREAD

**Issue by itsmeboy0** [rickbutton/workspacer#212]
Some of my friends hate the bar sitting on the top of the screen, so I looked at [win3wm](https://github.com/McYoloSwagHam/win3wm) which has a less intrusive status bar on top of the taskbar.
![status bar](https://github.com/McYoloSwagHam/win3wm/raw/master/images/showcase_1.png)
Might be good to implement in workspacer as an alternative of the top bar.

**Comment by dalyIsaac** [comment]
Just to clarify, you mean a bar at the bottom which has a limited width?

**Comment by lauverityhill** [comment]
> Just to clarify, you mean a bar at the bottom which has a limited width?

I believe he's talking about the Title Bars being removed just like on the top right cmd window.
![image](https://user-images.githubusercontent.com/68158964/123542305-8f4bf080-d77b-11eb-9f61-b98b6a3ba796.png)

**Comment by system32uwu** [comment]
Yeah I also think he's talking about the titlebars of the windows, in that case we can use the Win32 API, win3wm does this [here](https://github.com/McYoloSwagHam/win3wm/blob/abdc3362c387469b9d7d14804c2efe932affb6ec/guitest/wm.cpp#L1458).

**Comment by dalyIsaac** [comment]
Yeah, that seems like a good feature.

nog has something similar - https://github.com/TimUntersberger/nog/blob/master/assets/default_config.lua
nog also has `has_custom_titlebar` which allows special handling for some executables.

**Comment by schiltz3** [comment]
I played around with removing the Title Bar, and if someone does implement this. I would heavily recommend doing it in a way that supports special handling for some executables because messing with the GWL_STYLE is very inconsistent.

**Comment by itsmeboy0** [comment]
> Just to clarify, you mean a bar at the bottom which has a limited width?

Actually, I was talking about this. An alternative to the black bar sitting on the top of the screen.

**Comment by dalyIsaac** [comment]
I've created #217 to track hiding the title bars - we can use this issue to track @itsmeboy0's point.

**Comment by lauverityhill** [comment]
> Actually, I was talking about this. An alternative to the black bar sitting on the top of the screen.

Sorry, my bad

**Comment by dalyIsaac** [comment]
Nah it raised a good point

**Comment by sitiom** [comment]
> ![status bar](https://github.com/McYoloSwagHam/win3wm/raw/master/images/showcase_1.png)

Makes sense. It would be nice to keep using the taskbar without an extra bar taking up the free space.

**Comment by N1x0** [comment]
I'm testing [here](https://github.com/workspacer/workspacer/commit/102139f2bf6bc9ef8a63f98ea3cf823c846851ea), will submit a pull request when it's ready.

My current approach is to add two new properties `MaxWidth` and `BarIsTop` to `BarPluginConfig`. These are then used in `BarForm` to calculate the position of the bar. Lastly the `MenuBarLayoutEngine` has to be adjust the offsets for the window calculations depending on whether the bar is on top or at the bottom

E.g.

**workspacer.config.csx**

```
 context.AddBar(new BarPluginConfig()
    {
        BarTitle = "workspacer.Bar",
        BarHeight = 30,
        FontSize = 12,
        DefaultWidgetForeground = Color.Black,
        DefaultWidgetBackground = Color.White,
        BarMaxWidth = 1000,
        BarIsTop = false,
    
        LeftWidgets = () => new IBarWidget[] { new WorkspaceWidget() { WorkspaceHasFocusColor = Color.Blue } },
        RightWidgets = () => new IBarWidget[] { }
    });
````

**BarForm.csx**
```
        if (_config.BarIsTop)
            this.Location = new Point(_monitor.X, _monitor.Y - titleBarHeight);
        else

            this.Location = new Point(_monitor.X, _monitor.Height - _config.BarHeight);```
```

**MenuBarLayoutEngine**

```
      if (_barIsTop)
            {
                return _inner.CalcLayout(newWindows, spaceWidth, spaceHeight - _offset)
                .Select(l => new WindowLocation(l.X, l.Y + _offset, l.Width, l.Height, l.State));
            }
            else
            {
                return _inner.CalcLayout(newWindows, spaceWidth, spaceHeight - _offset)
                .Select(l => new WindowLocation(l.X, l.Y, l.Width, l.Height, l.State));
            }
        }
```

**Result**
![image](https://user-images.githubusercontent.com/47623143/123677878-d4514f00-d86f-11eb-8e75-c8f4a4f1bf3b.png)


Things to consider:
1. If the bar is at the bottom it should move when the windows taskbar auto-hides, this will require an event listener
2. If we do implement other taskbar positions already does it make sense to also cover left/right alignment?

**Comment by sitiom** [comment]
Since it's just a bar with a limited width attached to the taskbar, it should not take up the entire horizontal space on the bottom. Instead, it should be laid out on top of the workspace (screenshot from win3wm):

![image](https://user-images.githubusercontent.com/56180050/123730982-e233be00-d8c9-11eb-8ec2-117c914cbcc4.png)

**Comment by N1x0** [comment]
> Since it's just a bar with a limited width attached to the taskbar, it should not take up the entire horizontal space on the bottom. Instead, it should be laid out on top of the workspace (screenshot from win3wm):
> 
> ![image](https://user-images.githubusercontent.com/56180050/123730982-e233be00-d8c9-11eb-8ec2-117c914cbcc4.png)
> 

Just to confirm my understanding is correct, you'd suggest that if the bar is in bottom position, it should offset the window position using `MenuBarLayoutEngine`? If so, that could be handled by a flag or set as default behaviour. Easy to do

**Comment by sitiom** [comment]
> > Since it's just a bar with a limited width attached to the taskbar, it should not take up the entire horizontal space on the bottom. Instead, it should be laid out on top of the workspace (screenshot from win3wm):
> > ![image](https://user-images.githubusercontent.com/56180050/123730982-e233be00-d8c9-11eb-8ec2-117c914cbcc4.png)
> 
> Just to confirm my understanding is correct, you'd suggest that if the bar is in bottom position, it should **not** offset the window position using `MenuBarLayoutEngine`? If so, that could be handled by a flag or set as default behaviour. Easy to do

Yep, might make sense to set as default behavior.

> I'm testing [here](https://github.com/workspacer/workspacer/commit/102139f2bf6bc9ef8a63f98ea3cf823c846851ea), will submit a pull request when it's ready.

Also, you could open it now as a Draft PR if you want to do so.

**Comment by N1x0** [comment]
Created draft PR #222, please comment on implementation scope and features you'd deem necessary
